src/celeritas/mat/TabulatedElementSelector.hh

```diff
@@ -0,0 +1,102 @@
+ *
+ * This selects an elemental component (atom) of a material based on the
+ * precalculated cross section CDF tables of the elements in the material.
+ * Unlike \c ElementSelector which calculates the microscopic cross sections on
+ * the fly, this interpolates the values using tabulated CDF grids.
+ */
+class TabulatedElementSelector
+{
+  public:
+    //!@{
+    //! Type aliases
+    using Energy = Quantity<XsGridData::EnergyUnits>;
+    using GridValues
+        = Collection<ValueGrid, Ownership::const_reference, MemSpace::native>;
+    using GridIdValues
+        = Collection<ValueGridId, Ownership::const_reference, MemSpace::native>;
+    using Values
+        = Collection<real_type, Ownership::const_reference, MemSpace::native>;
+    //!@}
+
+  public:
+    // Construct with xs CDF data for a particular model and material
+    inline CELER_FUNCTION TabulatedElementSelector(const ValueTable&   table,
+                                                   const GridValues&   grids,
+                                                   const GridIdValues& ids,
+                                                   const Values&       reals,
+                                                   Energy              energy);
+
+    // Sample with the given RNG
+    template<class Engine>
+    inline CELER_FUNCTION ElementComponentId operator()(Engine& rng) const;
+
+  private:
+    const ValueTable&   table_;
+    const GridValues&   grids_;
+    const GridIdValues& ids_;
+    const Values&       reals_;
+    const Energy        energy_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with xs CDF data for a particular model and material.
+ */
+CELER_FUNCTION
+TabulatedElementSelector::TabulatedElementSelector(const ValueTable&   table,
+                                                   const GridValues&   grids,
+                                                   const GridIdValues& ids,
+                                                   const Values&       reals,
+                                                   Energy              energy)
+    : table_(table), grids_(grids), ids_(ids), reals_(reals), energy_(energy)
+{
+    CELER_EXPECT(table);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample the element with the given RNG.
+ */
+template<class Engine>
+CELER_FUNCTION ElementComponentId
+TabulatedElementSelector::operator()(Engine& rng) const
+{
+    size_type i = 0;
+    real_type u = generate_canonical(rng);
+    for (; i < table_.grids.size(); ++i)
+    {
+        ValueGridId grid_id = ids_[table_.grids[i]];
+        CELER_ASSERT(grid_id < grids_.size());
+        XsCalculator calc_xs(grids_[grid_id], reals_);
+        if (calc_xs(energy_) > u)
+            break;
+    }
+    return ElementComponentId{i};
```

Comment:
Good eyes -- and in fact I did run into an off-the-end error when trying this branch out with full CMS last night (though for another reason: there seem to be cases where the micro xs are all zero). I used a linear search to start since the grids are quite small, but that's a good point, I'll add it to the list.